src/Analyser/NodeScopeResolver.php

```diff
@@ -955,8 +967,10 @@ private function processStmtNode(
 					$bodyScope = $bodyScope->mergeWith($continueExitPoint->getScope());
 				}
 				foreach ($stmt->loop as $loopExpr) {
-					$bodyScope = $this->processExprNode($loopExpr, $bodyScope, static function (): void {
-					}, ExpressionContext::createTopLevel())->getScope();
+					$exprResult = $this->processExprNode($loopExpr, $bodyScope, static function (): void {
+					}, ExpressionContext::createTopLevel());
+					$bodyScope = $exprResult->getScope();
+					$hasYield = $hasYield || $exprResult->hasYield();
```

Comment:
not sure why this is a special case - i guess we could also get the `ExpressionResult` before this loop on `$stmt->loop`